File: backend/app/schemas/dataset.py
"""
Dataset-related Pydantic schemas
"""

from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime

from .common import BboxModel, StyleClassification

# Dataset schemas
class DatasetBase(BaseModel):
    """Base dataset fields"""
    name: str = Field(..., min_length=1, max_length=255, description="Dataset name")
    description: Optional[str] = Field(None, description="Dataset description")
    tags: List[str] = Field(default_factory=list, description="Dataset tags")

class DatasetCreate(DatasetBase):
    """Create dataset request"""
    source: str = Field(..., description="Dataset source: upload, huggingface, url")
    source_url: Optional[str] = Field(None, description="Source URL if applicable")

class Dataset(DatasetBase):
    """Dataset response model"""
    id: str = Field(..., description="Dataset ID")
    source: str = Field(..., description="Dataset source")
    source_url: Optional[str] = Field(None, description="Source URL")
    total_scenes: int = Field(..., description="Total number of scenes")
    processed_scenes: int = Field(..., description="Number of processed scenes")
    total_objects: int = Field(..., description="Total number of detected objects")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")

    class Config:
        from_attributes = True

# Scene schemas
class SceneObjectBase(BaseModel):
    """Base scene object fields"""
    label: str = Field(..., description="Object label")
    confidence: float = Field(..., ge=0, le=1, description="Detection confidence")
    bbox: BboxModel = Field(..., description="Bounding box")
    category: Optional[str] = Field(None, description="Object category")
    material: Optional[str] = Field(None, description="Material classification")
    material_conf: Optional[float] = Field(None, ge=0, le=1, description="Material confidence")
    description: Optional[str] = Field(None, description="Object description")

class SceneObject(SceneObjectBase):
    """Scene object response model"""
    id: str = Field(..., description="Object ID")
    scene_id: str = Field(..., description="Scene ID")
    r2_key_mask: Optional[str] = Field(None, description="R2 key for segmentation mask")
    r2_key_thumbnail: Optional[str] = Field(None, description="R2 key for thumbnail")
    review_status: Optional[str] = Field(None, description="Review status")
    review_notes: Optional[str] = Field(None, description="Review notes")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")

    class Config:
        from_attributes = True

class SceneBase(BaseModel):
    """Base scene fields"""
    source: str = Field(..., description="Original filename")
    width: int = Field(..., gt=0, description="Image width")
    height: int = Field(..., gt=0, description="Image height")

class SceneCreate(SceneBase):
    """Create scene request"""
    r2_key_original: str = Field(..., description="R2 key for original image")
    attrs: Optional[Dict[str, Any]] = Field(None, description="Additional attributes")

class Scene(SceneBase):
    """Scene response model"""
    id: str = Field(..., description="Scene ID")
    dataset_id: str = Field(..., description="Dataset ID")
    dataset_name: Optional[str] = Field(None, description="Dataset name")
    r2_key_original: str = Field(..., description="R2 key for original image")
    r2_key_thumbnail: Optional[str] = Field(None, description="R2 key for thumbnail")
    r2_key_depth: Optional[str] = Field(None, description="R2 key for depth map")
    
    # Classification results
    scene_type: Optional[str] = Field(None, description="Scene type classification")
    scene_conf: Optional[float] = Field(None, ge=0, le=1, description="Scene classification confidence")
    styles: List[StyleClassification] = Field(default_factory=list, description="Style classifications")
    palette: List[str] = Field(default_factory=list, description="Color palette (hex colors)")
    
    # Status
    status: str = Field(..., description="Processing status")
    processing_error: Optional[str] = Field(None, description="Processing error message")
    review_status: Optional[str] = Field(None, description="Review status")
    review_notes: Optional[str] = Field(None, description="Review notes")
    reviewed_by: Optional[str] = Field(None, description="Reviewer name")
    reviewed_at: Optional[datetime] = Field(None, description="Review timestamp")
    
    # Objects (optional, loaded on demand)
    objects: Optional[List[SceneObject]] = Field(None, description="Detected objects")
    
    # Timestamps
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")
    processed_at: Optional[datetime] = Field(None, description="Processing completion timestamp")

    class Config:
        from_attributes = True

# Upload-related schemas
class PresignFileRequest(BaseModel):
    """Single file presign request"""
    filename: str = Field(..., description="Original filename")
    content_type: str = Field(..., description="File MIME type")

class PresignRequest(BaseModel):
    """Presigned URLs request"""
    files: List[PresignFileRequest] = Field(..., description="Files to upload")

class PresignUpload(BaseModel):
    """Single presigned upload response"""
    filename: str = Field(..., description="Original filename")
    key: str = Field(..., description="R2 object key")
    url: str = Field(..., description="Presigned upload URL")
    headers: Dict[str, str] = Field(..., description="Required headers")

class PresignResponse(BaseModel):
    """Presigned URLs response"""
    uploads: List[PresignUpload] = Field(..., description="Upload information")

class RegisterScenesRequest(BaseModel):
    """Register scenes request"""
    scenes: List[SceneCreate] = Field(..., description="Scenes to register")

class RegisterScenesResponse(BaseModel):
    """Register scenes response"""
    created: int = Field(..., description="Number of scenes created")
    scene_ids: List[str] = Field(..., description="Created scene IDs")

File: react-app/src/types/dataset.ts
export interface Dataset {
  id: string
  name: string
  version: string
  source_url?: string
  license?: string
  notes?: string
  created_at: string
}

export interface DatasetCreate {
  name: string
  version: string
  source_url?: string
  license?: string
  notes?: string
}

export interface Scene {
  id: string
  dataset_id: string
  dataset_name?: string
  source: string
  r2_key_original: string
  r2_key_thumbnail?: string
  r2_key_depth?: string
  width: number
  height: number
  scene_type?: string
  scene_conf?: number
  styles?: StyleClassification[]
  palette?: ColorPalette[]
  has_depth: boolean
  objects_count: number
  objects?: SceneObject[]
  review_status?: ReviewStatus
  review_notes?: string
  reviewed_by?: string
  reviewed_at?: string
  created_at: string
  updated_at?: string
}

export interface SceneObject {
  id: string
  scene_id: string
  label: string
  confidence: number
  bbox: BoundingBox
  r2_key_mask?: string
  material?: string  // Legacy single material
  material_conf?: number  // Legacy material confidence
  materials?: MaterialDetection[]  // Enhanced: multiple materials with confidence
  primary_material?: string  // Enhanced: top material
  material_confidence?: number  // Enhanced: top material confidence
  review_status?: ReviewStatus
  corrected_label?: string
  review_notes?: string
}

export interface MaterialDetection {
  material: string
  confidence: number
}

export interface BoundingBox {
  x: number
  y: number
  width: number
  height: number
}

export type ReviewStatus = 'pending' | 'approved' | 'rejected' | 'corrected'

export interface SceneReview {
  scene_id: string
  reviewer_id: string
  status: ReviewStatus
  corrected_scene_type?: string
  corrected_styles?: StyleClassification[]
  notes?: string
  created_at: string
}

export interface ObjectReview {
  object_id: string
  reviewer_id: string
  status: ReviewStatus
  corrected_label?: string
  corrected_material?: string
  notes?: string
  created_at: string
}

export interface ReviewSession {
  id: string
  reviewer_id: string
  dataset_id?: string
  scenes_reviewed: number
  time_spent_seconds: number
  started_at: string
  ended_at?: string
}

export interface StyleClassification {
  code: string
  conf: number
}

export interface ColorPalette {
  hex: string
  p: number
}

export interface Job {
  id: string
  kind: 'ingest' | 'process'
  status: 'queued' | 'running' | 'succeeded' | 'failed' | 'skipped'
  dataset_id: string
  dataset_name?: string
  started_at?: string
  finished_at?: string
  progress?: JobProgress
  error_message?: string
  error_trace?: string
  created_at: string
  updated_at: string
}

export interface JobProgress {
  scenes_total: number
  scenes_done: number
  objects_detected: number
  avg_conf: number
  failures: number
  current_scene?: string
  eta_seconds?: number
  processing_rate?: number
}

export interface JobCreate {
  kind: 'ingest' | 'process'
  dataset_id: string
  config?: JobConfig
}

export interface JobConfig {
  scene_classifier_threshold?: number
  style_classifier_threshold?: number
  object_detector_threshold?: number
  material_detector_threshold?: number
  force_reprocess?: boolean
}

export interface JobLogEntry {
  timestamp: string
  level: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR'
  message: string
  context?: Record<string, any>
}

export interface JobStats {
  total_jobs: number
  queued_jobs: number
  running_jobs: number
  completed_jobs: number
  failed_jobs: number
  average_processing_time: number
  success_rate: number
}

export interface PresignRequest {
  files: PresignFile[]
}

export interface PresignFile {
  filename: string
  content_type: string
}

export interface PresignResponse {
  uploads: PresignUpload[]
  expires_in_seconds: number
}

export interface PresignUpload {
  filename: string
  key: string
  url: string
  headers: Record<string, string>
}

export interface RegisterScenesRequest {
  scenes: SceneRegistration[]
}

export interface SceneRegistration {
  source: string
  r2_key_original: string
  width: number
  height: number
}

export interface RegisterScenesResponse {
  created: number
  scene_ids: string[]
}

export interface Page<T> {
  items: T[]
  page: number
  limit: number
  total: number
}

export interface APIError {
  error: {
    code: string
    message: string
    trace_id: string
  }
}

// Roboflow API types
export interface ProcessRoboflowRequest {
  roboflow_url: string
  api_key: string
  export_format?: string
  max_images?: number
}

export interface ProcessRoboflowResponse {
  job_id: string
  status: string
}

export interface RoboflowDatasetInfo {
  workspace: string
  project: string
  version: string
  dataset_id: string
  description: string
  tags: string[]
  license: string
  format: string
}